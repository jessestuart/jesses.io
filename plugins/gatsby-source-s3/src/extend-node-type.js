/* @flow */
const Promise = require('bluebird')
const fs = require('fs')
const {
  GraphQLObjectType,
  GraphQLString,
  GraphQLInt,
  GraphQLFloat,
} = require('graphql')
const exif = require('exif-parser')
const DateTime = require('luxon').DateTime
const _ = require('lodash')

/**
 * ExifData persists the exif data parsed from an image binary
 * within Gatsby's GraphQL data infra. The best way to access
 * these fields is directly via the S3ImageAsset node -- e.g.,
 * @example
 * ```graphql
 * {
 *   allS3ImageAsset {
 *     edges {
 *       node {
 *         id
 *         EXIF {
 *           DateCreatedISO
 *           FNumber
 *           // ...etc
 *         }
 *       }
 *     }
 *   }
 * }
 * ```
 *
 * Note that you can also obtain direct access to the `ImageSharp`
 * node as a child relation:
 * @example
 * ```graphql
 * {
 *   allS3ImageAsset {
 *     edges {
 *       node {
 *         id
 *         EXIF {
 *           DateCreatedISO
 *           FNumber
 *           // ...etc
 *         }
 *       }
 *     }
 *   }
 * }
 * ```
 */
type ExifData = {
  DateCreatedISO: String,
  DateTimeOriginal: Number,
  ExposureTime: Number,
  FNumber: Number,
  FocalLength: Number,
  ISO: Number,
  LensModel: String,
  Model: String,
  ShutterSpeedValue: Number,
}

/**
 * S3ImageAssetNode is a minimal wrapper composing the default Node
 * fields with those obtained from S3 -- initially just `Key`
 * and `ETag` (object digest), but the full map of Exif
 * properties are injected during the `extend-node-type` step.
 */
type S3ImageAssetNode = {
  id: String,
  absolutePath: String,
  ETag: String,
  Key: String,
  EXIF: ?ExifData,
  internal: {
    content: String,
    contentDigest: String,
    mediaType: String,
    type: String,
  },
}

const resolveExifData = (image: S3ImageAssetNode): ExifData => {
  const file = fs.readFileSync(image.absolutePath)
  const tags = exif.create(file).parse().tags
  const timestamp = tags.DateTimeOriginal * 1000
  const DateCreatedISO = DateTime.fromMillis(timestamp).toISODate()
  return {
    DateCreatedISO,
    ..._.pick(tags, [
      'DateTimeOriginal',
      'ExposureTime',
      'FNumber',
      'FocalLength',
      'ISO',
      'LensModel',
      'Model',
      'ShutterSpeedValue',
    ]),
  }
}

type ExtendNodeTypeOptions = {
  type: {
    name: String,
  },
}

const extendNodeType = ({ type }: ExtendNodeTypeOptions) => {
  if (type.name !== 'S3ImageAsset') {
    return {}
  }

  return Promise.resolve({
    ETag: { type: GraphQLString },
    Key: { type: GraphQLString },
    EXIF: {
      type: new GraphQLObjectType({
        name: 'ExifData',
        fields: {
          DateCreatedISO: { type: GraphQLString },
          DateTimeOriginal: { type: GraphQLInt },
          ExposureTime: { type: GraphQLFloat },
          FNumber: { type: GraphQLFloat },
          FocalLength: { type: GraphQLFloat },
          ISO: { type: GraphQLInt },
          LensModel: { type: GraphQLString },
          Model: { type: GraphQLString },
          ShutterSpeedValue: { type: GraphQLFloat },
        },
      }),
      resolve(image: S3ImageAssetNode) {
        return {
          ...type,
          ...resolveExifData(image),
        }
      },
    },
  })
}

exports.extendNodeType = extendNodeType

exports.resolveExifData = resolveExifData
