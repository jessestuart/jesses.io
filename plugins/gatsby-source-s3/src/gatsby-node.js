exports.sourceNodes = require('./source-nodes').sourceNodes

exports.setFieldsOnGraphQLNodeType = require(`./extend-node-type`).extendNodeType
