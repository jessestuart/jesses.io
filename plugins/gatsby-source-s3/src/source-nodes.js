// =========================
// Third party dependencies.
// =========================
const _ = require('lodash')
const AWS = require('aws-sdk')
const Promise = require('bluebird')

// ============
// Gatsby APIs.
// ============
const { createRemoteFileNode } = require('gatsby-source-filesystem')

// =================
// AWS config setup.
// =================
const S3 = new AWS.S3({ apiVersion: '2006-03-01' })

// =========================
// Plugin-specific contants.
// =========================
const S3SourceGatsbyNodeType = 'S3ImageAsset'

// =================
// Type definitions.
// =================
type SourceS3Options = {
  bucketName: String,
  // Defaults to `${bucketName}.s3.amazonaws.com`, but may be
  // overridden to e.g., support CDN's (such as CloudFront),
  // or any other S3-compliant API (such as DigitalOcean
  // Spaces.)
  domain: ?String,
  // Defaults to HTTPS.
  protocol: ?String,
}

const constructS3UrlForAsset = ({
  bucketName,
  domain,
  key,
  protocol = 'https',
}): ?String => {
  // Both `key` and either one of `bucketName` or `domain` are required.
  if (!key || (!bucketName && !domain)) {
    return null
  }
  // If `domain` is defined, that takes precedence over `bucketName.`
  if (domain) {
    return `${protocol}://${domain}/${key}`
  }
  if (bucketName) {
    return `${protocol}://${bucketName}.s3.amazonaws.com/${key}`
  }
}

exports.sourceNodes = async (
  { boundActionCreators, getNode, hasNodeChanged, store, cache },
  { bucketName, domain, protocol }: SourceS3Options,
  done
) => {
  const { createNode } = boundActionCreators

  const listObjectsResponse = await S3.makeUnauthenticatedRequest(
    'listObjectsV2',
    {
      Bucket: bucketName,
    }
  ).promise()
  const s3Entities = _.get(listObjectsResponse, 'Contents')

  await Promise.all(
    s3Entities.map(async entity => {
      const s3Url = constructS3UrlForAsset({
        bucketName,
        domain,
        key: entity.Key,
        protocol,
      })
      const entityData = {
        bucketName,
        cache,
        createNode,
        domain,
        entity,
        protocol,
        store,
        s3Url,
      }

      const fileNode = await createS3RemoteFileNode(entityData)
      if (fileNode) {
        entityData.localFile___NODE = fileNode.id
      }
      await createS3ImageAssetNode({
        ...entityData,
        fileNode,
        done,
      })
    })
  )
  done()
}

const createS3RemoteFileNode = async ({ cache, createNode, store, s3Url }) => {
  try {
    return await createRemoteFileNode({
      url: s3Url,
      store,
      cache,
      createNode,
    })
  } catch (err) {
    // eslint-disable-next-line
    console.error('Unable to create file node.', err)
    return null
  }
}

const createS3ImageAssetNode = async ({
  createNode,
  done,
  entity,
  fileNode,
  s3Url,
}) => {
  const { Key, ETag } = entity
  // TODO: Could probably pull this from fileNode.
  const ContentType = 'image/jpeg'
  // Remove obnoxious escaped double quotes in S3 object's ETag. For reference:
  // > The entity tag is a hash of the object. The ETag reflects changes only
  // > to the contents of an object, not its metadata.
  // @see https://docs.aws.amazon.com/AmazonS3/latest/API/RESTCommonResponseHeaders.html
  const objectHash = ETag.replace(/"/g, '')
  createNode({
    ...entity,
    id: `${Key} >> ${S3SourceGatsbyNodeType}`,
    absolutePath: fileNode.absolutePath,
    Key,
    parent: fileNode.id,
    children: [],
    internal: {
      content: s3Url,
      contentDigest: objectHash,
      mediaType: ContentType,
      type: S3SourceGatsbyNodeType,
    },
  })
}
