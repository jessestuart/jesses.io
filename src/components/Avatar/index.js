import React from 'react'

import './styles.scss'

export default function Avatar() {
  const isDev = process.env.NODE_ENV === 'development'
  const pathPrefix = isDev ? '' : __PATH_PREFIX__
  return (
    <img className="js-avatar" src={pathPrefix + '/img/avatar_square.jpg'} />
  )
}
