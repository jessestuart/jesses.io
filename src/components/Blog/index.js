export { default as BlogHeader } from './BlogHeader'
export { default as BlogNoPosts } from './BlogNoPosts'
