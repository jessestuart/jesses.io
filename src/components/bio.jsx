import React from 'react'

import { Follow } from 'react-twitter-widgets'
import { rhythm } from '../utils/typography'
import './bio.scss'

function year() {
  const d = new Date()
  return d.getFullYear()
}

class Bio extends React.Component {
  render() {
    return (
      <div style={{ display: 'flex', justifyContent: 'space-around', marginBottom: rhythm(2.5) }}>
        <div className='follow'>
          <Follow username='cwpittman' options={{ size: 'medium', dnt: true }} />
        </div>
        <div>
          Â© <strong>Cameron Pittman</strong> {year()}.
        </div>
      </div>
    )
  }
}

export default Bio
