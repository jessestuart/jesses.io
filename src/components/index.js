/* eslint-disable */
export Layout from './layout'
export Profile from './Profile/Profile'
export ProfileDevIcons from './ProfileDevIcons'
export SiteFooter from './SiteFooter/SiteFooter'
export SiteHeader from './SiteHeader/SiteHeader'
export StyledLink from './StyledLink/StyledLink'
export StyledPanel from './StyledPanel/StyledPanel'
