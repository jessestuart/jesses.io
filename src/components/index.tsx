import { PortfolioItem } from './Portfolio'
import BlogHeader from './Blog/BlogHeader'
import HeaderLink from './SiteHeader/HeaderLink'
import ProfileDevIcons from './ProfileDevIcons'
import SiteFooter from './SiteFooter/SiteFooter'
import SiteHeader from './SiteHeader/SiteHeader'
import StyledImg from './StyledImg'
import StyledLink from './StyledLink/StyledLink'
import StyledPanel from './StyledPanel/StyledPanel'

export * from './Photography'
export * from './Profile'

export {
  // Avatar,
  // BlackMirror,
  BlogHeader,
  HeaderLink,
  // ImageZoomGrid,
  // PhotographyGridSection,
  // PhotographySectionHeader,
  PortfolioItem,
  // Profile,
  // ProfileBlurb,
  ProfileDevIcons,
  // ProfileFooter,
  // ProfileHeading,
  // ProfileStyledLink,
  SiteFooter,
  SiteHeader,
  StyledLink,
  StyledPanel,
  StyledImg,
}
