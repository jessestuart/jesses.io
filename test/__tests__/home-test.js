import React from 'react'

import renderer from 'react-test-renderer'

import { Avatar } from '../../src/components'

it('renders correctly', () => {
  const tree = renderer.create(<Avatar />).toJSON()
  expect(tree).toMatchSnapshot()
  // const div = document.createElement('div')
  // ReactDOM.render(<Avatar />, div)
})
