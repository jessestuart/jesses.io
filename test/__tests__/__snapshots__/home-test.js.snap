// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<img
  className="br-100 flex justify-center mv2 pa0 shadow-6"
  src="test-file-stub"
/>
`;
